### DISCUSSION TITLE
How to setup interactive books using Java

### DISCUSSION THREAD

**Discussion by cagix** [LiaScript/LiaScript#156] [Q&A] [answered]
I would like to use interactive executable Java examples in my interactive books/presentations. Can you give me some tips on how to do this? I've searched the documentation and there are even examples for C and Prolog, but I'm not clear on how to integrate interactive Java examples. Thx!

**Comment by SebastianZug** [comment, marked as answer]
Hi Carsten,
take a view on [LiaLiveExample](https://liascript.github.io/LiveEditor/?/show/code/H4sIAAAAAAAAA3WQT2sCMRDF7/Mpxu3Fhe7u3ZZSsYIWe9FDDyLsmI1rSv6RZNdK8bs3uypYobkkvPfy+M08D7IMhLLGhRHuQ7B+VBSODnktwr7ZNp47ZnTgOuTMqGIhaMWcsKGYmIovG625KxT5EK/ldPz2Mc1VBVn2AvCA79QSTr9JWcnxEPuQxU+u/5QDzDtdxWYKwmjk56DHnXFYid2Ou+ihdaZ2pJTQNUrSdUN1zDDSuOUx2ugKjQa4op+x/2UFgLIs8asjQ5xxKc2ncbLKOwVss5WCIZPk/Y2JPxDDeHF9x8uwNaJCRUIPV8FFuPUGydU+vYS7szrGvajcNCG3MRKkHiZ9az+iOmLUuCMWRMv7ZWUT0zjPB0n61Jec4NThwutiPs55S3JYrpM76GRTPmLZvdj9PFfjRi9TgBlZe/yz11GW/gLuXcnGCAIAAA== ). I built a 'HelloWorld' example for Java.
Best wishes.
Sebastian

**Reply by cagix** [reply]
Thanks for this example, it looks really promising!

Am I right that this example is running server-side? I.e. if I were to embed it in my lecture notes and the students were to "download" it as a PWA, it would stop working, right?

Is there a way to define the JDK version and possibly use additional libraries?

**Reply by SebastianZug** [reply]
Hi,
no, the "format" of a liascript course does not affect coderunner's functionality. But an internet connection is required to reach our instance at TU Freiberg. Feel free to deploy your own service - a short explanation is given in the last section of the [README.md](https://liascript.github.io/course/?https://raw.githubusercontent.com/liascript/CodeRunner/master/README.md#25).
Sebastian

**Reply by Zireael07** [reply]
The link goes to a description of Golang language in Lia, not to any sort of a deployment instruction

**Reply by andre-dietrich** [reply]
@Zireael07 this was just the example for Go, if you open the table of contents, then there are more languages that can be embedded ... the actual implementation is defined here:

https://github.com/LiaScript/CodeRunner/blob/master/Dockerfile

you can install your own languages and host it by your own, it is just an Ubuntu with lots of languages preinstalled ... to change the path to your installation, simply clone the repo or at least the [README.md](https://raw.githubusercontent.com/LiaScript/CodeRunner/refs/heads/master/README.md), which need to be included and change the link to your deployment:

`window.CodeRunner.init("wss://your-instance.io")`

**Reply by andre-dietrich** [reply]
@cagix I just updated @SebastianZug s example, if you have only one file/class, then there is a shortcut, to which you only need to pass the name of the class and all other steps are defined automatically. If you need more files and more control, you can use the `@LIA.eval` option, take a look at this [Live Example](https://liascript.github.io/LiveEditor/?/show/code/H4sIAAAAAAAAA61TTW/aQBC976+YmItpwag9krZqRJFClUhVOOQQRfKyXsxW+2HtBwRV+e+dNQaMReFSLmtm37x5M/P2y81wSISqjPVjWHlfufFoZOkmK4VfhUVw3DKjPdc+Y0aNHgSdMysqP5qYgj8FrbkdKeo8Hk/Tux+P00wVZDj8RkgPftI1hekbVZXksEE+YJhk66SMkFmMK2SmXhgNfAd0sDQWCrFccot3UFlTWqqU0CVIqstAS8QwqmHBERp0AUYTspe+k/1PrYSQPM/hd1QGcM+lNM/GyiKLEVKFhRQMmKTOtS7hD0EwNLcu6mWwNqIARYVO596iuJdXoLZ0/QYcf/MtzkVlJvisQoiXOk1q1rpFtQWMcUuZF2teD2s4McE6fpP0b2uSd/Ie5ZLvD7O7WmF6FNUn5J5W1fZkQOMhhns9eDSWwyS2wV2r4wmVLEjqjT3T7/HytF8UCbQo0voc1H8X7S4t98FqoPARFnvZ3XwVpBeV3F4n+XAk2fV+EN9d1s609V0WvJDZHE2BK74l/32PO2JwzfkVNN/so2mzZaFxaddXP4BaymHF58BpMo3GAL9ChwvrPOigFtyOoZ0VZ4jxTyinEZZp/uZnmH9JSZvccXzbxSX2z1fYW6Zh+NmM5hhNO5QuKMREaBY9FfUP6jodHJq6CMzvsQf/nCRcnPa8rpSgK7HmhWFH7K9DtYhvarcL7EfApHE8Pfc6+ZrKNH9JOk8sGUDSMW7ymg8gj1+s6+nuA90DW7gcH/hfbONbjbQFAAA=)

----

````` markdown
<!--
import: https://raw.githubusercontent.com/LiaScript/CodeRunner/master/README.md
-->

# Java Example with coderunner.

Implementation examples for different programming languages can be found on

https://github.com/LiaScript/CodeRunner


``` java   HelloWorld.java
public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello for my interactive Java-Course!");
    }
}
```
@LIA.java(HelloWorld)

Happy programming :-)

## More Classes

``` java Calculator.java
public class Calculator {
    public int add(int a, int b) {
        return a + b;
    }

    public int multiply(int a, int b) {
        return a * b;
    }
}
```
``` java HelloWorld.java
import java.util.Scanner;

public class HelloWorld {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Hello, World!");
        System.out.print("Enter the first number: ");
        int num1 = scanner.nextInt();

        System.out.print("Enter the second number: ");
        int num2 = scanner.nextInt();

        Calculator calc = new Calculator();
        int sum = calc.add(num1, num2);
        int product = calc.multiply(num1, num2);

        System.out.println("Sum = " + sum);
        System.out.println("Product = " + product);

        scanner.close();
    }
}
```
@LIA.eval(`["Calculator.java", "HelloWorld.java"]`, `javac HelloWorld.java Calculator.java`, `java HelloWorld`)
`````

you define a list of filenames as the first parameter in the order of code-snippets, then the compilation command and finally the execution ...

**Comment by andre-dietrich** [comment]
If you need some special libraries, they can be added here:

https://github.com/LiaScript/CodeRunner/blob/3382620c7d60e169758062cbc0146f07b63ed95c/Dockerfile#L30

You can either create a pull-request with your libraries, or simply tell us what you would like to add ;-)

**Reply by cagix** [reply]
Thanks a lot! As far as the libraries are concerned, I don't have anything specific in mind yet, but who knows what will come up in the preparation phase. However, I already know that I will need a JDK 21 (LTS).

**Reply by andre-dietrich** [reply]
we are currently using JDK 24 ... is this a problem?

https://github.com/LiaScript/CodeRunner/blob/a3f2b8bbbdd2ff6e3816729adcdb20ac033dd32f/Dockerfile#L139-L142

**Reply by cagix** [reply]
hmmm, not really. usually i'm using the current lts version, but for the next semester i'm planning on switching to java 25 (better switch/case, pattern matching, streams, ... compared to java 21).
